### ISSUE TITLE
Disabling node in cluster stuck

### ISSUE THREAD

**Issue by alecor191** [Azure/service-fabric-issues#328]
I'm trying to remove a node from a node type. I'm following the documentation in [Manually remove VMs from the primary node type/VM scale set](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-cluster-scale-up-down#manually-remove-vms-from-the-primary-node-typevm-scale-set).

Unfortunately I'm already stuck at the first step: Calling `Disable-ServiceFabricNode` triggers the disable process, however, it seems to be stuck in that state: After 25min, `Get-ServiceFabricNode -NodeName _tmp2-nt0_1` still shows the following:

```text
NodeName             : _tmp2-nt0_1
NodeId               : e9c5e713b990505911e0423e294d73ab
NodeInstanceId       : 131427781133848489
NodeType             : nt0
NodeStatus           : Disabling
HealthState          : Ok
CodeVersion          : 5.6.210.9494
ConfigVersion        : 2
IsSeedNode           : True
IpAddressOrFQDN      : 10.0.0.5
FaultDomain          : fd:/1
UpgradeDomain        : 1
NodeDeactivationInfo : EffectiveIntent : RemoveNode
                       Status : SafetyCheckInProgress

                       TaskType : Client
                       TaskId : e9c5e713b990505911e0423e294d73ab
                       Intent : RemoveNode

                       PendingSafetyChecks :
                        EnsureSeedNodeQuorum
IsStopped            : False
```

![screenshot1](https://user-images.githubusercontent.com/6270008/27508427-e706299a-58e4-11e7-86aa-1f845f767803.png)

![screenshot2](https://user-images.githubusercontent.com/6270008/27508428-ec026f9e-58e4-11e7-8caf-8a5cf1926bb8.png)

After an hour I got impatient and used SF Explorer's interface to invoke action "Restart" on that node. It didn't seem to have much effect, as the node is still stuck in "Disabling". The only difference is that now I also get the health metrics (Source: "System.FM" and "System.FMM", property "State") showing messages "**Fabric node is taking longer than expected to deactivate.**"

### Cluster information
- Single node type, 6 nodes, DS2_v2 (2 cores, 7GB memory), 'bronze' reliability tier (i.e. minimum of 3 nodes required)
- Azure-hosted cluster, version 5.6.210.9494
- At the time the above was performed, a single app was deployed to the cluster (5 stateless services) and all nodes were idle.
- Note that I'm trying to delete a node that is _not_ the last one (i.e. _don't_ do the default scale in/out scenario where the last nodes are removed; I want to simulate a node in the VMSS getting bad and thus I want to remove it from the cluster).

Any thoughts on what could be the reason for disabling a node being stuck?

**Comment by dolly22** [comment]
You have bronze reliability with 3 seed nodes (very special nodes _0, _1, _2). You are trying to remove one of them which would mean you cluster would probably loose seed node vote quorum (notice your pending `EnsureSeedNodeQuorum` safety check)

> Indicates that if we bring down the node then this will result in global seed node quorum loss.

I suggest you don't try to remove your seed nodes, unless you want to encounter some problems (https://github.com/Azure/service-fabric-issues/issues/185). I have not yet found a way to recreate concrete seed node in virtual machine scale set.

The only information about service fabric seed nodes i have found...

From FederationConfig documentation (azure clusters use SeedNode configuration type)
> A vote represents a single count towards a quorum of in a cluster. A vote is assigned to a vote owner(node) by a vote authority. Vote Authorities are typically nodes (seed nodes) in the cluster. A cluster needs a quorum of votes present to guarantee its health and ability to stay up.  Losing a quorum of votes will cause the cluster to go down. An alternative to seed nodes are SQL votes; where the vote authority is not a node in the Cluster; but a SQL Server instance. In this case; the node with the closest Id to the SQL vote Id acts as a proxy. The votes selected through configuration; must be the same across all nodes. ID is a string that is parsed into a long integer that represents the ID of the vote in the cluster. The Type can be one of two options: SeedNode or SqlServer; depending on the vote authority used by the cluster. Depending on the type is the format for the connection string. The connection string for SeedNode is the NodeEndpoint for the node with the same NodeID. For type=SqlServer is a connection string to a SQL Server 2008 and up. An example to a SeedNode is : '0 = SeedNode;10.0.0.1:10000'. An example to a SQL vote is: 'sqlvote1 = SqlServer;Provider=SQLNCLI10;Server=.\SQLEXPRESS;Database=master;Integrated Security=SSPI'

From [Azure Service Fabric MSDN forum](https://social.msdn.microsoft.com/Forums/en-US/a4b9e755-d2e5-4f32-916d-56da04ba6467/how-does-service-fabric-handle-a-tie-when-there-is-an-even-number-of-nodes?forum=AzureServiceFabric)
> You're confusing service replicas with voting. Voting rights are given to specially-designated nodes called seed nodes. This is, as Karol says, an implementation detail of Service Fabric's clustering technology. Service replicas provide high-availability of services and their state where quorum refers to the majority required to respond to a replication operation for it to be considered complete.

**Comment by alecor191** [comment]
Thanks @dolly22 for your input! My expectation was that removing a seed node would trigger SF to elect one of the non-seed nodes to become a seed node.

The scenario I want to make sure works with SF is that if *any* VM in the VMSS dies, it can be taken out and replaced with another one.

If we can never remove seed nodes, it means that sooner or later the cluster will go down, as we need to be prepared that any VM will go down eventually (updates, OS upgrade, HW failure,...).

I see that your last reply in #185 pretty much represents the scenario I'm testing. Looking forward to see feedback to your questions from MS to understand how to deal with this in production environments.

**Comment by rishirsinha** [comment]
Can you look at the node health for the other seed nodes? It might just be that while the VM is up something is busted with one of the seed nodes. Not that I am saying things o.w. are perfectly fine and there is not an issue at all, but Bronze is not supported in production for this very reason. Once a seed node is stuck you can not upgrade your way out at all.

**Comment by alecor191** [comment]
Thanks @rishirsinha. We did a couple more tests with various cluster sizes and durability/reliability tiers:

With silver durability we observed that on a VMSS with DS2v2 nodes, taking down a *non-seed* node took ~10min (incl. disabling node in SF and then removing the node from the VMSS - disabling the node in SF was a couple seconds; bulk of time spent in VMSS). 

Removing a *seed* node took ~1.5h where most of the time (>1h) was spent in SF to disable the node. So it seems in my initial investigation above I was a bit impatient. After the 1.5h we could observe that one of the *non-seed* nodes got promoted to *seed* node, ensuring there is no seed node quorum loss.

To summarize: The scenario of removing seed nodes/non-seed nodes seems to work just fine (at least with durability silver), it just takes longer than we expected.
